qutip/solver/heom/bofin_baths.py

```diff
@@ -337,6 +346,181 @@ def combine(cls, exponents, rtol=1e-5, atol=1e-7):
+        function according to the list of expansion coefficients that is used
+        for the HEOM construction.
+
+        Note that various conventions for the definition of the spectral
+        density exist. We follow the convention described in Eq. 7 of the BoFiN
+        paper (DOI: 10.1103/PhysRevResearch.5.013181). We also assume that
+        the spectral density is zero at negative frequencies (otherwise, no
+        thermal state exists).
+        """
+
+        raise NotImplementedError(
+            "The spectral density of this bath has not been specified")
+
+    def correlation_function(self, t, **kwargs):
+        """
+        Calculates the correlation function from the spectral density by
+        numeric integration, see Eq. 6 in the BoFiN paper
+        (DOI: 10.1103/PhysRevResearch.5.013181). This calculation requires that
+        the temperature of the bath (`self.T`) is specified.
+
+        Parameters
+        ----------
+        t : np.array or float
+            the time(s) at which to evaluate the correlation function
+        kwargs : will be passed to scipy's `quad_vec` function
+
+        Returns
+        -------
+            The correlation function at time t as an array or float
+        """
+
+        def integrand(w, t):
+            return self.spectral_density(w) / np.pi * (
+                (2 * self._bose_einstein(w) + 1) * np.cos(w * t)
+                - 1j * np.sin(w * t)
+            )
+
+        result = quad_vec(lambda w: integrand(w, t), 0, np.Inf, **kwargs)
+        return result[0]
+
+    def _bose_einstein(self, w):
+        """
+        Calculates the bose einstein distribution for the
+        temperature of the bath.
+
+        Parameters
+        ----------
+        w: float or array
+            Energy of the mode
+
+        Returns
+        -------
+        The population of the mode with energy w
+        """
+
+        if self.T is None:
+            raise NotImplementedError(
+                "Bath temperature must be specified for this operation")
+        if self.T == 0:
+            return np.zeros_like(w)
+
+        w = np.array(w, dtype=float)
+        with np.errstate(divide='ignore'):  # return inf if w=0
+            return (1 / (np.exp(w / self.T) - 1))
+
+    def power_spectrum(self, w):
+        """
+        Calculates the power spectrum from the spectral density using the
+        fluctuation dissipation theorem. This calculation requires that the
+        temperature of the bath (`self.T`) is specified.
+
+        Parameters
+        ----------
+        w: float or array
+            Energy of the mode
+
+        Returns
+        ----------
+        The power spectrum of the mode with energy w
+        """
+
+        # For w=0, the result would have the form 0 * inf
+        # We shift w by an epsilon to get the value in the limit w->0
+        w = np.array(w, dtype=float)
+        w[w == 0.0] += 1e-6
+
+        S = (2 * np.sign(w) * self.spectral_density(np.abs(w)) *
```

Comment:
It's not clear to me where the `inf` would come from in the `0 * inf` mentioned in the comment. Possibly from `_bose_einstein`? But see the question higher up that points out that `_bose_einstein` currently returns zeroes when `T = 0`.

In any case, it would be nice to handle this better if we can.

Reply:
The Bose-Einstein factor at w=0 diverges (see above). At the same time, the spectral density at w=0 is zero.

More precisely, the Bose-Einstein factor diverges as 1/w. Therefore, the power spectrum at w=0 can be zero, finite or infinite depending on the behavior of the SD at small w (subohmic, ohmic or superohmic). Without that line 446, we would return NaN at w=0. The best idea we could come up with was evaluating the expression somewhere close to zero instead. If you have a better idea, we would be happy to hear it.

(Note that this function is only a utility function for the purpose of visualization etc, so it might be not too important to get this 100% exact.)